auth-service/src/test/kotlin/uk/gov/communities/delta/security/OAuthLoginTest.kt

```diff
@@ -0,0 +1,326 @@
+import uk.gov.communities.delta.auth.plugins.configureTemplating
+import uk.gov.communities.delta.auth.security.azureAdSingleSignOn
+import uk.gov.communities.delta.auth.security.configureRateLimiting
+import uk.gov.communities.delta.auth.services.AuthCode
+import uk.gov.communities.delta.auth.services.AuthorizationCodeService
+import uk.gov.communities.delta.auth.services.UserLookupService
+import uk.gov.communities.delta.auth.services.sso.MicrosoftGraphService
+import uk.gov.communities.delta.auth.services.sso.SSOLoginSessionStateService
+import uk.gov.communities.delta.auth.services.sso.SSOOAuthClientProviderLookupService
+import uk.gov.communities.delta.helper.testLdapUser
+import uk.gov.communities.delta.helper.testServiceClient
+import java.time.Instant
+import javax.naming.NameNotFoundException
+import kotlin.test.assertContains
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertTrue
+
+
+class OAuthLoginTest {
+
+    @Test
+    fun testOAuthFlow() = testSuspend {
+        // OAuth flow happy path
+        val testClient = testClient()
+
+        // Get the login page (required as it sets a cookie with the Delta client id and state in)
+        testClient.get("/delta/login?response_type=code&client_id=delta-website&state=delta-state").apply {
+            assertEquals(HttpStatusCode.OK, status)
+            assertContains(bodyAsText(), "Test SSO</a>")
+        }
+
+        // Request to the OAuth login endpoint as though we'd pressed the SSO button
+        val oauthLoginResponse = testClient.get("/delta/oauth/test/login").apply {
+            assertEquals(HttpStatusCode.Found, status)
+        }
+        // Expect a redirect to Microsoft's Authorize endpoint
+        val externalServiceRedirect = oauthLoginResponse.headers["Location"]
+        assertNotNull(externalServiceRedirect, "Location header expected")
+        assertTrue(externalServiceRedirect.startsWith("https://login.microsoftonline.com/${ssoClient.azTenantId}/oauth2/v2.0/authorize"))
+        assertContains(externalServiceRedirect, "client_id=${ssoClient.azClientId}")
+        assertContains(
+            externalServiceRedirect,
+            "redirect_uri=${"http://auth-service/delta/oauth/test/callback".encodeURLParameter()}"
+        )
+        val state = externalServiceRedirect.stateFromRedirectUrl()
+
+        // Microsoft would then redirect the user back to the redirect ("callback") endpoint
+        testClient.get("/delta/oauth/test/callback?code=auth-code&state=${state}").apply {
+            // Which should redirect us back to Delta with an Authorisation code
+            assertEquals(HttpStatusCode.Found, status)
+            assertEquals("https://delta/login/oauth2/redirect?code=code&state=delta-state", headers["Location"])
+            verify { authorizationCodeServiceMock.generateAndStore("cn", serviceClient, any()) }
+            assertEquals("", setCookie()[0].value) // Session should be cleared
```

Comment:
Why does this use setCookie not getCookie or is it just called setCookie?

Reply:
It's called setCookie because the response header is "Set-Cookie" i.e. it's asking the browser to set a cookie. So it is getting the cookies from the response.